lookout/style/format/features.py

```diff
@@ -607,46 +781,66 @@ def _inplace_write_vnode_features(
                 # NOOP. If the current node is NOOP, then its direct neighbours are interesting
                 # (read non-NOOP) nodes.
                 start_left = i - 1
-                end_left = start_left - self.siblings_window * 2
-                start_right = i + 1
-                end_right = start_right + self.siblings_window * 2
+                end_left = start_left - self.left_siblings_window * 2
                 # For non-NOOP nodes, the first interesting nodes are further from the current node
                 # since there are NOOP nodes in-between.
                 if vnode.y != CLASS_INDEX[CLS_NOOP]:
                     start_left -= 1
                     end_left -= 1
-                    start_right += 1
-                    end_right += 1
                 # We go two by two to avoid NOOP nodes.
                 left_siblings = vnodes[max(start_left, 0):max(end_left, 0):-2]
-                right_siblings = vnodes[start_right:end_right:2]
+
+                # If we don't compute right siblings without labeled nodes, we do the same for
+                # them.
+                if not self.no_labels_on_right:
+                    start_right = i + 1
+                    end_right = start_right + self.right_siblings_window * 2
+                    if vnode.y != CLASS_INDEX[CLS_NOOP]:
+                        start_right += 1
+                        end_right += 1
+                    right_siblings = vnodes[start_right:end_right:2]
             else:
-                left_siblings = vnodes[max(i - 1, 0):max(i - self.siblings_window - 1, 0):-1]
-                right_siblings = vnodes[i + 1:i + self.siblings_window + 1]
+                left_siblings = vnodes[max(i - 1, 0):max(i - self.left_siblings_window - 1, 0):-1]
+                right_siblings = vnodes[i + 1:i + self.right_siblings_window + 1]
+
+            if self.no_labels_on_right:
+                right_siblings = []
+                for j in range(i + 1, len(vnodes)):
+                    if len(right_siblings) >= self.right_siblings_window:
+                        break
+                    if vnodes[j].y is None:
+                        right_siblings.append(vnodes[j])
 
             col_offset = 0
             # 1. write features of the current node
-            col_offset += self._inplace_write_features(self._get_self_features(vnode),
-                                                       position, col_offset, X)
+            col_offset += self._inplace_write_features(
+                self._get_features(self.node_features, vnode, None), position, col_offset, X)
```

Comment:
function header:
```
def _get_features(self, features, sibling, node):
    ...
```
I think you mean `self._get_features(self.node_features, None, vnode)` or different order in the header.

Reply:
No, it's the correct order, it's just that here the "sibling" and the "central node" are the same. We can probably do more elaborate signatures for Feature to avoid having such cases but I'd rather do it after demo.

Reply:
sure